File: blind75/51_group_anagrams/group_anagrams.go
package group_anagrams

import (
	"sort"
	"strings"
)

// GroupAnagrams groups anagrams together from the input array.
// Uses hash map with sorted string as key for O(n*k*log(k)) time complexity.
func GroupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)

	for _, s := range strs {
		key := sortString(s)
		groups[key] = append(groups[key], s)
	}

	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}

	return result
}

func sortString(s string) string {
	chars := strings.Split(s, "")
	sort.Strings(chars)
	return strings.Join(chars, "")
}

File: blind75/51_group_anagrams/group_anagrams.py
from typing import List
from collections import defaultdict


def group_anagrams(strs: List[str]) -> List[List[str]]:
    """
    Groups anagrams together from the input array.
    Uses hash map with sorted string as key for O(n*k*log(k)) time complexity.

    Args:
        strs: List of strings to group

    Returns:
        List of lists, where each inner list contains anagrams
    """
    groups = defaultdict(list)

    for s in strs:
        key = "".join(sorted(s))
        groups[key].append(s)

    return list(groups.values())
